src/bsdfs/bumpmap.cpp

```diff
@@ -0,0 +1,210 @@
+            if (bsdf) {
+                if (m_nested_bsdf)
+                    Throw("Only a single BSDF child object can be specified.");
+                m_nested_bsdf = bsdf;
+                props.mark_queried(name);
+            }
+            auto texture = dynamic_cast<Texture *>(obj.get());
+            if (texture) {
+                if (m_nested_texture)
+                    Throw("Only a single Texture child object can be specified.");
+                m_nested_texture = texture;
+                props.mark_queried(name);
+            }
+        }
+        if (!m_nested_bsdf)
+            Throw("Exactly one BSDF child object must be specified.");
+        if (!m_nested_texture)
+            Throw("Exactly one Texture child object must be specified.");
+
+        m_scale = props.float_("scale", 1.f);
+
+        // Add all nested components
+        m_components.clear();
+        for (size_t i = 0; i < m_nested_bsdf->component_count(); ++i)
+            m_components.push_back(m_nested_bsdf->flags(i));
+        m_flags = m_nested_bsdf->flags();
+    }
+
+    std::pair<BSDFSample3f, Spectrum> sample(const BSDFContext &ctx,
+                                             const SurfaceInteraction3f &si,
+                                             Float sample1,
+                                             const Point2f &sample2,
+                                             Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::BSDFSample, active);
+
+        // Sample nested BSDF with perturbed shading frame
+        SurfaceInteraction3f perturbed_si(si);
+        perturbed_si.sh_frame = frame(si, active);
+        perturbed_si.wi = perturbed_si.to_local(si.wi);
+        auto [bs, weight] = m_nested_bsdf->sample(ctx, perturbed_si,
+                                                  sample1, sample2, active);
+        active &= any(neq(depolarize(weight), 0.f));
+        if (none(active))
+            return { bs, 0.f };
+
+        // Transform sampled 'wo' back to original frame and check orientation
+        Vector3f perturbed_wo = perturbed_si.to_world(bs.wo);
+        active &= Frame3f::cos_theta(bs.wo) *
+                  Frame3f::cos_theta(perturbed_wo) > 0.f;
+        bs.wo = perturbed_wo;
+
+        return { bs, weight & active };
+    }
+
+    Spectrum eval(const BSDFContext &ctx, const SurfaceInteraction3f &si,
+                  const Vector3f &wo, Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::BSDFEvaluate, active);
+
+        // Evaluate nested BSDF with perturbed shading frame
+        SurfaceInteraction3f perturbed_si(si);
+        perturbed_si.sh_frame = frame(si, active);
+        perturbed_si.wi       = perturbed_si.to_local(si.wi);
+        Vector3f perturbed_wo = perturbed_si.to_local(wo);
+
+        active &= Frame3f::cos_theta(wo) *
+                  Frame3f::cos_theta(perturbed_wo) > 0.f;
+
+        return m_nested_bsdf->eval(ctx, perturbed_si, perturbed_wo, active);
```

Comment:
I think that we at some point need some better documentation on masking (my fault). The policy is that a function is allowed to return any garbage when the mask flag is `false`. (@Speierers, we may want to revisit this policy at some point because it introduces inefficiencies).  For now, you will want to wrap this in another `select(active, ..., 0)`.